File: heron/examples/src/python/word_count_topology.py
# copyright 2016 twitter. all rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from heron.examples.src.python.word_spout import WordSpout
from heron.examples.src.python.count_bolt import CountBolt
from heron.instance.src.python.basics import Topology, Grouping

import heron.common.src.python.constants as constants

class WordCount(Topology):
  task_hooks = ["heron.examples.src.python.test_task_hook.TestTaskHook"]
  config = {constants.TOPOLOGY_ENABLE_ACKING: "true",
            constants.TOPOLOGY_MAX_SPOUT_PENDING: 100000000,
            constants.TOPOLOGY_AUTO_TASK_HOOKS: task_hooks,
            "topology.wide.config.sample": {"key1": 12, "key2": 34}}

  word_spout = WordSpout.spec(par=1)
  count_bolt = CountBolt.spec(par=1,
                              inputs={word_spout: Grouping.fields('word'),
                                      word_spout['error']: Grouping.ALL},
                              config={constants.TOPOLOGY_TICK_TUPLE_FREQ_SECS: 10,
                                      "count_bolt.specific": ["123", (12, 34)]})

File: storm-compatibility/tests/java/org/apache/storm/WordCountTopology.java
// Copyright 2016 Twitter. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.storm;

import org.apache.storm.generated.AlreadyAliveException;
import org.apache.storm.generated.InvalidTopologyException;
import org.apache.storm.generated.NotAliveException;
import org.apache.storm.topology.TopologyBuilder;
import org.apache.storm.tuple.Fields;

import com.twitter.heron.common.basics.ByteAmount;

public class WordCountTopology {

  public static final ByteAmount RAM_IN_BYTES = ByteAmount.fromMegabytes(16);
  private TopologyBuilder builder;
  private Config conf;

  @SuppressWarnings("deprecation")
  public WordCountTopology() throws InterruptedException, InvalidTopologyException,
      NotAliveException,
      AlreadyAliveException {
    builder = new TopologyBuilder();

    builder.setSpout("sentence", new RandomSentenceSpout(), 1);
    builder.setBolt("split", new SplitSentenceBolt(), 1).shuffleGrouping("sentence");
    builder.setBolt("count", new WordCountBolt(), 1)
        .fieldsGrouping("split", new Fields("word"));
    conf = new Config();

    // Resource Configs
    com.twitter.heron.api.Config.setComponentRam(conf, "sentence", RAM_IN_BYTES);
    com.twitter.heron.api.Config.setComponentRam(conf, "split", RAM_IN_BYTES);
    com.twitter.heron.api.Config.setComponentRam(conf, "count", RAM_IN_BYTES);
    com.twitter.heron.api.Config.setContainerCpuRequested(conf, 1);

  }

  public void execute() throws InterruptedException, InvalidTopologyException, NotAliveException,
      AlreadyAliveException {
    Util.runTopologyLocally(builder.createTopology(), "WordCountTopology", conf, 10);
  }

}

